File: web/packages/core/src/providers/relay-form-validation.ts
/**
 * Declarative field definitions plus pure validation for the custom LLM (relay) form.
 *
 * Why this file exists: the create page (`RelaySetup.tsx`) and the edit page (`RelayDetail.tsx` /
 * `RelayConnectionCard`) each carried their own answer to "can this be saved", and the same rule
 * drifted three ways: edit required a non-empty key regardless of authMode, create required one too,
 * and edit always measured the address against the default `remote_https` mode, so a local engine
 * `http://` address was permanently invalid and even renaming it could not be saved.
 *
 * There is exactly one place per rule: endpoint validity is delegated to `classifyRelayEndpoint` (the
 * same classifier the send boundary uses) and credential requirements to
 * `relayCredentialInputRequired`. This file only owns the form-level field layout and combination
 * rules; it reimplements none of the underlying checks.
 *
 * Not done here: merging the create and edit pages into one component. Their state machines differ
 * fundamentally (create has probe-generation guards, one-token verification and an empty-catalog exit;
 * edit has credential reuse and catalog invalidation), and merging would just smear the two together.
 */
import type { RelayAuthMode, RelayKeyValue, RelayTransport } from '@oriveo/shared/pure-types';
import {
  classifyRelayEndpoint,
  isSensitiveRelayName,
  type RelayConnectionSecurityMode,
} from '@oriveo/shared/relay/endpoint-policy';
import { isPrintableAsciiKey } from './key-validation';
import { relayCredentialInputRequired, relayRequiresCredential } from './relay-runtime-support';

/* ── Field layout (declarative definitions) ───────────── */

/**
 * Form fields, one per field of `RelayFormDraft`.
 *
 * The display name is not among them: in the create flow it is an optional display label and in the
 * edit flow it goes through a separate rename entry point, so it takes part in no validation.
 */
export type RelayFormField =
  | 'endpoint'
  | 'security_mode'
  | 'api_key'
  | 'model_id'
  | 'transport'
  | 'auth_mode'
  | 'headers'
  | 'query_params';

/** Field groups. The order is the order the groups appear in the form. */
export type RelayFormGroup = 'connection' | 'protocol' | 'advanced_http';

/** Whether editing this field invalidates the current probe/verification generation. */
export type RelayFormRevalidation =
  | 'always'
  | 'never'
  /**
   * Default model only: switching to another entry within an already discovered catalog does not
   * invalidate the probe (the protocol was verified against that same catalog, and clearing the result
   * would make the catalog picker itself disappear). Typing an ID from outside the catalog does mean the
   * user changed target.
   */
  | 'unless_within_discovered_catalog';

export interface RelayFormFieldDefinition {
  readonly field: RelayFormField;
  readonly group: RelayFormGroup;
  /** Full next-intl key for the label. The connection mode is a read-only summary row on both pages. */
  readonly labelKey: string | null;
  /**
   * Placeholder. Address, key and KV are non-localized example literals; the default model
   * placeholder is dynamic (the flagship model ID published in metadata) and is therefore `null`.
   */
  readonly placeholder: string | null;
  readonly revalidation: RelayFormRevalidation;
}

/**
 * Golden layout order: the array order is the top-to-bottom render order, used by both the create and
 * edit pages. `security_mode` comes after the address, so the address is described first and how to
 * reach it second.
 */
export const RELAY_FORM_FIELDS: readonly RelayFormFieldDefinition[] = [
  {
    field: 'endpoint',
    group: 'connection',
    labelKey: 'pages.relayDetail.endpoint',
    placeholder: 'https://api.example.com/v1',
    revalidation: 'always',
  },
  {
    field: 'security_mode',
    group: 'connection',
    labelKey: 'pages.relayDetail.connectionType',
    placeholder: null,
    revalidation: 'always',
  },
  {
    field: 'api_key',
    group: 'connection',
    labelKey: 'pages.providerDetail.apiKey',
    placeholder: 'sk-...',
    revalidation: 'always',
  },
  {
    field: 'model_id',
    group: 'connection',
    labelKey: 'pages.relaySetup.defaultModelLabel',
    placeholder: null,
    revalidation: 'unless_within_discovered_catalog',
  },
  {
    field: 'transport',
    group: 'protocol',
    labelKey: 'pages.relayDetail.transport',
    placeholder: null,
    revalidation: 'always',
  },
  {
    field: 'auth_mode',
    group: 'protocol',
    labelKey: 'pages.relayDetail.authMode',
    placeholder: null,
    revalidation: 'always',
  },
  {
    field: 'headers',
    group: 'advanced_http',
    labelKey: 'pages.relayDetail.headers',
    placeholder: 'X-Internal-Token',
    revalidation: 'always',
  },
  {
    field: 'query_params',
    group: 'advanced_http',
    labelKey: 'pages.relayDetail.queryParams',
    placeholder: 'tenant',
    revalidation: 'always',
  },
];

export function relayFormFieldDefinition(
  field: RelayFormField,
): RelayFormFieldDefinition | undefined {
  return RELAY_FORM_FIELDS.find((definition) => definition.field === field);
}

/* ── Validation results ───────────────────────────────── */

export type RelayFormIssueCode =
  /** The address is empty. */
  | 'endpoint_required'
  /** The address was rejected by the address policy; `detail` holds its reason slug. */
  | 'endpoint_rejected'
  /** Credential material (key, sensitive header or sensitive query) sent over a cleartext transport. */
  | 'cleartext_credentials'
  /** Connection mode and address scheme are incompatible. The mode is never auto-adjusted to fit the address. */
  | 'security_mode_scheme_mismatch'
  /** This connection needs a key and the form has none yet. */
  | 'credential_required'
  /** The key contains non-printable ASCII, most often full-width or zero-width spaces pasted along with it. */
  | 'credential_invalid_characters';

export interface RelayFormIssue {
  readonly field: RelayFormField;
  readonly code: RelayFormIssueCode;
  /** Machine-readable detail (for a rejected endpoint, the address policy's reason slug). Not shown in the UI. */
  readonly detail?: string;
}

/**
 * "Required field" issues only keep the primary button disabled: an empty field is its own hint, and
 * red text on first paint is noise. Everything else means the user filled something in incorrectly and
 * needs a plain-language message right away, otherwise the button stays grey with no explanation.
 */
export function isSilentRelayFormRequirement(code: RelayFormIssueCode): boolean {
  return code === 'endpoint_required' || code === 'credential_required';
}

/** The issues that warrant an immediate plain-language message. */
export function displayableRelayFormIssues(
  issues: readonly RelayFormIssue[],
): RelayFormIssue[] {
  return issues.filter((issue) => !isSilentRelayFormRequirement(issue.code));
}

/* ── Draft state and validation ───────────────────────── */

export type RelayFormMode = 'create' | 'edit';

/** Form draft. The create and edit pages keep their own state machines but both reduce state into this draft for validation. */
export interface RelayFormDraft {
  endpoint: string;
  apiKey: string;
  authMode: RelayAuthMode;
  securityMode: RelayConnectionSecurityMode;
  transport: RelayTransport;
  modelID: string;
  headers: readonly RelayKeyValue[];
  queryParams: readonly RelayKeyValue[];
  /** A non-empty key is already in the keystore. Always `false` in `create`, which has no keystore. */
  hasSavedCredential: boolean;
}

const FIELD_ORDER: readonly RelayFormField[] = RELAY_FORM_FIELDS.map((item) => item.field);

/**
 * Pure form validation. No side effects, no network requests, no global state reads.
 *
 * The returned order always matches the layout order of `RELAY_FORM_FIELDS`, and one field can return
 * several issues (address first, then credentials).
 */
export function validateRelayForm(
  draft: RelayFormDraft,
  mode: RelayFormMode,
): RelayFormIssue[] {
  const issues: RelayFormIssue[] = [];
  const trimmedEndpoint = draft.endpoint.trim();
  const trimmedKey = draft.apiKey.trim();
  // In `edit` the keystore already holds a key: leaving the input empty means "unchanged", not "clear".
  // In `create` there is no keystore, only the key the user just typed.
  const hasCredential = trimmedKey.length > 0 || (mode === 'edit' && draft.hasSavedCredential);

  // 1. Address
  if (trimmedEndpoint.length === 0) {
    issues.push({ field: 'endpoint', code: 'endpoint_required' });
  } else {
    if (hasSchemeModeMismatch(trimmedEndpoint, draft.securityMode)) {
      issues.push({ field: 'security_mode', code: 'security_mode_scheme_mismatch' });
    }
    // The same classifier the send boundary uses, so form validation and saving cannot disagree.
    const classification = classifyRelayEndpoint({
      raw: trimmedEndpoint,
      securityMode: draft.securityMode,
      credentials: {
        authMode: draft.authMode,
        hasKey: hasCredential,
        sensitiveHeaders: sensitiveNames(draft.headers),
        sensitiveQueryKeys: sensitiveNames(draft.queryParams),
      },
    });
    if (!classification.allowed) {
      if (classification.reason === 'cleartext_credentials') {
        issues.push({
          field: cleartextOffendingField(draft, hasCredential),
          code: 'cleartext_credentials',
          detail: classification.reason,
        });
      } else {
        issues.push({
          field: 'endpoint',
          code: 'endpoint_rejected',
          detail: classification.reason,
        });
      }
    }
  }

  // 2. Credential requirement (the single rule lives in the credential state machine; no thresholds are copied here)
  if (relayCredentialInputRequired(mode, draft.authMode, hasCredential)) {
    issues.push({ field: 'api_key', code: 'credential_required' });
  }

  // 3. Key character set. An empty string is valid for `auth=none` and for leaving the field blank while
  //    editing, so only non-empty keys are checked.
  if (trimmedKey.length > 0 && !isPrintableAsciiKey(trimmedKey)) {
    issues.push({ field: 'api_key', code: 'credential_invalid_characters' });
  }

  return issues
    .map((issue, index) => ({ issue, index }))
    .sort((lhs, rhs) => {
      const delta = fieldIndex(lhs.issue.field) - fieldIndex(rhs.issue.field);
      return delta !== 0 ? delta : lhs.index - rhs.index;
    })
    .map((entry) => entry.issue);
}

/** Normalized address once validation passes, `null` otherwise. Both paths share one check rather than recomputing. */
export function relayFormNormalizedEndpoint(
  draft: RelayFormDraft,
  mode: RelayFormMode,
): string | null {
  if (validateRelayForm(draft, mode).length > 0) return null;
  const hasCredential = draft.apiKey.trim().length > 0
    || (mode === 'edit' && draft.hasSavedCredential);
  const classification = classifyRelayEndpoint({
    raw: draft.endpoint.trim(),
    securityMode: draft.securityMode,
    credentials: {
      authMode: draft.authMode,
      hasKey: hasCredential,
      sensitiveHeaders: sensitiveNames(draft.headers),
      sensitiveQueryKeys: sensitiveNames(draft.queryParams),
    },
  });
  return classification.allowed ? classification.normalized ?? null : null;
}

/* ── Internal checks ──────────────────────────────────── */

function fieldIndex(field: RelayFormField): number {
  const index = FIELD_ORDER.indexOf(field);
  return index < 0 ? FIELD_ORDER.length : index;
}

function sensitiveNames(pairs: readonly RelayKeyValue[]): string[] {
  return pairs.filter((pair) => isSensitiveRelayName(pair.key)).map((pair) => pair.key);
}

/**
 * Consistency between the connection mode and the address scheme, checked at form level rather than
 * only on submit.
 *
 * Only addresses where the user wrote a scheme explicitly are checked: without one the address policy
 * fills in a scheme matching the current mode, which is always compatible. The case of an explicit
 * `http://` under a mode that requires encryption is reported by the address policy as
 * `cleartext_not_allowed`; this covers only the half it cannot see, an encrypted address under a mode
 * that declares cleartext.
 */
function hasSchemeModeMismatch(
  endpoint: string,
  securityMode: RelayConnectionSecurityMode,
): boolean {
  if (securityMode !== 'local_http' && securityMode !== 'private_vpn') return false;
  return endpoint.toLowerCase().startsWith('https://');
}

/** Which piece of credential material crossed the cleartext boundary. Attributed in layout order, so the user gets something to click. */
function cleartextOffendingField(
  draft: RelayFormDraft,
  hasCredential: boolean,
): RelayFormField {
  if (relayRequiresCredential(draft.authMode) || hasCredential) return 'api_key';
  if (draft.headers.some((pair) => isSensitiveRelayName(pair.key))) return 'headers';
  return 'query_params';
}

File: android/app/src/main/java/ai/oriveo/community/core/provider/RelayFormValidation.kt
package ai.oriveo.community.core.provider

import androidx.annotation.StringRes
import ai.oriveo.community.R
import ai.oriveo.community.core.model.ProviderServiceError
import ai.oriveo.community.core.model.RelayAuthMode
import ai.oriveo.community.core.model.RelayConnectionSecurityMode
import ai.oriveo.community.core.model.RelayKeyValue
import ai.oriveo.community.core.model.RelayRequestedConfig
import ai.oriveo.community.core.model.RelayTransport
import ai.oriveo.community.core.model.hasStoredCredential
import ai.oriveo.community.core.model.requiresCredential

/**
 * A declarative definition of the custom-endpoint form, plus pure-function validation for it.
 *
 * Why this file exists: the add page and the edit sheet each used to carry their own answer to
 * "can this be saved", and the same rule drifted apart three times. The add page only checked
 * that the endpoint was non-empty and a key was present, so an `http://` address on the local
 * network could be saved under a `remote_https` configuration - it saved fine and then every
 * send was hard-rejected by the send-time policy. The edit sheet measured the address against a
 * default of `remote_https`, so a local engine on `http://` was judged invalid forever and the
 * user could not even rename the connection. Field order and labels drifted independently too.
 *
 * There may be exactly one home for each judgement: endpoint legality is delegated to
 * [RelayEndpointPolicy] (the save path calls the very same
 * [RelayEndpointPolicy.requireConfigured]), and credential state to
 * `RelayAuthMode.requiresCredential` and [hasStoredCredential]. This file owns only the
 * form-level layout and combination rules; it never reimplements an underlying predicate.
 *
 * Cross-platform cases are pinned in `shared/test-fixtures/relay/form-validation.v1.json`.
 *
 * What this deliberately does not do: merge the add page and the edit sheet into one ViewModel.
 * The two state machines are genuinely different - add has probe-generation guards, single-token
 * validation and an empty-directory exit, while edit has "leave blank to keep" and directory
 * invalidation - and merging them would only smear both into one.
 */
object RelayFormValidation {

    // Field layout, declared once.

    /**
     * A form field. One-to-one with the fields of [RelayFormDraft].
     *
     * The display name is not among them: in the add flow it is an optional label, in the edit
     * flow it goes through its own rename entry point, it takes part in no validation, and it is
     * therefore not part of "the same form".
     */
    enum class Field(val value: String) {
        Endpoint("endpoint"),
        SecurityMode("security_mode"),
        ApiKey("api_key"),
        ModelId("model_id"),
        Transport("transport"),
        AuthMode("auth_mode"),
        Headers("headers"),
        QueryParams("query_params"),
    }

    /** Field groups. The order here is the order the groups appear in the form. */
    enum class Group(val value: String) {
        /** Connection: address, connection mode, key, default model. */
        Connection("connection"),

        /** Protocol: the wire protocol and the authentication mode. */
        Protocol("protocol"),

        /** Advanced HTTP: custom headers and query parameters. */
        AdvancedHttp("advanced_http"),
    }

    /** Whether changing this field invalidates the current probe or validation round. */
    enum class RevalidationTrigger(val value: String) {
        /** Any change invalidates it: address, key, connection mode, protocol, authentication
         *  mode, headers and query parameters. */
        Always("always"),

        /** Never invalidates. */
        Never("never"),

        /**
         * Default model only: picking another entry from the directory that was already probed
         * does not invalidate anything - the protocol has been verified against that same
         * directory, and clearing the result would make the model picker itself disappear.
         * Typing an id that is not in the directory does mean the user changed target.
         */
        UnlessWithinDiscoveredCatalog("unless_within_discovered_catalog"),
    }

    data class FieldDefinition(
        val field: Field,
        val group: Group,
        /**
         * String resource for the label.
         *
         * `null` means the field is not part of the visible form. `securityMode` has its own
         * always-present read-only summary row, so it must reference the real connection-mode
         * resource rather than hiding behind a null.
         */
        @StringRes val labelRes: Int?,
        /**
         * Placeholder text. Address, key and key/value rows use unlocalised example literals;
         * the default-model placeholder is dynamic (it shows a flagship model id from the
         * catalog), so it is `null` here.
         */
        val placeholder: String?,
        val revalidation: RevalidationTrigger,
    )

    /**
     * The canonical layout order: this list is the top-to-bottom render order, and both the add
     * page and the edit sheet follow it. [Field.SecurityMode] sits directly after the address -
     * say what the address is first, then how to reach it.
     */
    val fields: List<FieldDefinition> = listOf(
        FieldDefinition(
            field = Field.Endpoint,
            group = Group.Connection,
            labelRes = R.string.relay_field_endpoint,
            placeholder = "https://api.example.com/v1",
            revalidation = RevalidationTrigger.Always,
        ),
        FieldDefinition(
            field = Field.SecurityMode,
            group = Group.Connection,
            labelRes = R.string.relay_security_mode_label,
            placeholder = null,
            revalidation = RevalidationTrigger.Always,
        ),
        FieldDefinition(
            field = Field.ApiKey,
            group = Group.Connection,
            labelRes = R.string.api_key,
            placeholder = "sk-...",
            revalidation = RevalidationTrigger.Always,
        ),
        FieldDefinition(
            field = Field.ModelId,
            group = Group.Connection,
            labelRes = R.string.relay_default_model_label,
            placeholder = null,
            revalidation = RevalidationTrigger.UnlessWithinDiscoveredCatalog,
        ),
        FieldDefinition(
            field = Field.Transport,
            group = Group.Protocol,
            labelRes = R.string.relay_advanced_transport,
            placeholder = null,
            revalidation = RevalidationTrigger.Always,
        ),
        FieldDefinition(
            field = Field.AuthMode,
            group = Group.Protocol,
            labelRes = R.string.relay_advanced_auth_mode,
            placeholder = null,
            revalidation = RevalidationTrigger.Always,
        ),
        FieldDefinition(
            field = Field.Headers,
            group = Group.AdvancedHttp,
            labelRes = R.string.relay_advanced_headers,
            placeholder = "X-Internal-Token",
            revalidation = RevalidationTrigger.Always,
        ),
        FieldDefinition(
            field = Field.QueryParams,
            group = Group.AdvancedHttp,
            labelRes = R.string.relay_advanced_query_params,
            placeholder = "tenant",
            revalidation = RevalidationTrigger.Always,
        ),
    )

    fun definition(field: Field): FieldDefinition? = fields.firstOrNull { it.field == field }

    // Validation results.

    enum class IssueCode(
        val value: String,
        @StringRes val messageRes: Int,
        /**
         * A "required" issue only keeps the primary button disabled. An empty field is already
         * the hint, and floating a red line on open is noise. Everything else means the user
         * has filled something in and got it wrong, which has to be explained on the spot -
         * otherwise the button is greyed out and they cannot tell why.
         */
        val isSilentRequirement: Boolean,
    ) {
        /** The address is empty. */
        EndpointRequired("endpoint_required", R.string.relay_no_endpoint_set, true),

        /** [RelayEndpointPolicy] rejected the address; `detail` carries its reason slug. */
        EndpointRejected("endpoint_rejected", R.string.relay_setup_invalid_endpoint_message, false),

        /** Credential material is being carried over a cleartext transport (a key, a sensitive
         *  header, or a sensitive query parameter). */
        CleartextCredentials("cleartext_credentials", R.string.relay_credentials_cleartext_blocked, false),

        /**
         * The connection mode is incompatible with the address scheme. The mode is deliberately
         * never auto-adjusted to fit the address; the user has to change it explicitly.
         */
        SecurityModeSchemeMismatch(
            "security_mode_scheme_mismatch",
            R.string.relay_security_mode_scheme_mismatch,
            false,
        ),

        /** This connection needs a key and the form has none at all. */
        CredentialRequired("credential_required", R.string.error_api_key_required, true),

        /** The key contains non-printable ASCII, most often a full-width space or a newline
         *  dragged in by copy and paste. */
        CredentialInvalidCharacters(
            "credential_invalid_characters",
            R.string.provider_api_key_illegal_chars,
            false,
        ),
    }

    data class FieldIssue(
        val field: Field,
        val code: IssueCode,
        /** Machine-readable extra detail; for a rejected endpoint this is
         *  [RelayEndpointPolicy]'s reason slug. Never shown in the UI. */
        val detail: String? = null,
    ) {
        /**
         * The message to show. A rejected address is broken down by [RelayEndpointPolicy]'s
         * reason slug: "the address has a query string embedded in it" and "this is not HTTPS"
         * call for two completely different actions, and answering both with "enter a valid
         * HTTPS address" hides a cause we already know.
         */
        @get:StringRes
        val messageRes: Int get() = when {
            code == IssueCode.EndpointRejected && detail == "embedded_query" ->
                R.string.relay_quick_embedded_query
            else -> code.messageRes
        }
    }

    /** Which flow the form is in. The only real difference is that the edit flow knows a key is
     *  already in the keystore. */
    enum class FormMode { Create, Edit }

    /**
     * Whether the save action may treat the API key as a required field.
     *
     * - [FormMode.Create]: required only when the auth mode `requiresCredential` and no
     *   credential exists yet. Otherwise creating an `auth=none` connection would be pushed off
     *   to some parallel entry point with different capabilities.
     * - [FormMode.Edit]: always `false`. Leaving the input blank means "do not change it", not
     *   "clear it", and editing a non-credential field must never be gated on retyping the key.
     *   A genuinely missing credential is reported by the credential state instead.
     */
    fun credentialInputRequired(
        mode: FormMode,
        authMode: RelayAuthMode?,
        hasStoredKey: Boolean,
    ): Boolean = when (mode) {
        FormMode.Create -> (authMode ?: RelayAuthMode.Auto).requiresCredential && !hasStoredKey
        FormMode.Edit -> false
    }

    // Validation.

    /**
     * The form validation pure function. No side effects, no network, no global state.
     *
     * The returned order always matches the layout order in [fields], and one field can produce
     * more than one issue (address first, then credential).
     */
    fun validate(draft: RelayFormDraft, mode: FormMode): List<FieldIssue> {
        val issues = mutableListOf<FieldIssue>()
        val trimmedEndpoint = draft.endpoint.trim()
        val trimmedKey = draft.apiKey.trim()
        // Edit knows a key is already in the keystore: a blank input means "unchanged", not
        // "clear it". Create has no keystore, only whatever the user just typed.
        val hasCredential = hasStoredCredential(trimmedKey) ||
            (mode == FormMode.Edit && draft.hasSavedCredential)

        // 1. Address.
        if (trimmedEndpoint.isEmpty()) {
            issues += FieldIssue(Field.Endpoint, IssueCode.EndpointRequired)
        } else {
            if (hasSchemeModeMismatch(trimmedEndpoint, draft.securityMode)) {
                issues += FieldIssue(Field.SecurityMode, IssueCode.SecurityModeSchemeMismatch)
            }
            // The save path calls this same requireConfigured, so form validation and saving
            // cannot reach different conclusions.
            runCatching {
                RelayEndpointPolicy.requireConfigured(
                    baseUrl = trimmedEndpoint,
                    securityMode = draft.securityMode,
                    credentials = RelayEndpointPolicy.credentialsOf(
                        requested = draft.requestedConfigForPolicy(),
                        hasKey = hasCredential,
                    ),
                )
            }.onFailure { error ->
                val reason = reasonOf(error)
                if (reason == "cleartext_credentials") {
                    issues += FieldIssue(
                        field = cleartextOffendingField(draft, hasCredential),
                        code = IssueCode.CleartextCredentials,
                        detail = reason,
                    )
                } else {
                    issues += FieldIssue(Field.Endpoint, IssueCode.EndpointRejected, reason)
                }
            }
        }

        // 2. Credential required. The only predicate lives in requiresCredential; no threshold
        // is copied here.
        if (credentialInputRequired(mode, draft.authMode, hasCredential)) {
            issues += FieldIssue(Field.ApiKey, IssueCode.CredentialRequired)
        }

        // 3. Key character set. An empty string is legal for auth=none and for leaving the edit
        // field blank, so only a non-empty value is checked.
        if (trimmedKey.isNotEmpty() && !ProviderKeyInput.isPrintableAsciiKey(trimmedKey)) {
            issues += FieldIssue(Field.ApiKey, IssueCode.CredentialInvalidCharacters)
        }

        val order = fields.map { it.field }
        return issues.sortedBy { order.indexOf(it.field).takeIf { index -> index >= 0 } ?: order.size }
    }

    /**
     * The normalised address, once validation passes. `null` when it does not - both paths share
     * the one classification instead of computing it twice, and no caller is handed an address
     * that merely looks usable to write into storage.
     */
    fun normalizedEndpoint(draft: RelayFormDraft, mode: FormMode): String? {
        if (validate(draft, mode).isNotEmpty()) return null
        val hasCredential = hasStoredCredential(draft.apiKey.trim()) ||
            (mode == FormMode.Edit && draft.hasSavedCredential)
        return runCatching {
            RelayEndpointPolicy.requireConfigured(
                baseUrl = draft.endpoint.trim(),
                securityMode = draft.securityMode,
                credentials = RelayEndpointPolicy.credentialsOf(
                    requested = draft.requestedConfigForPolicy(),
                    hasKey = hasCredential,
                ),
            )
        }.getOrNull()
    }

    /** The issues that deserve an explanation on the spot (see [IssueCode.isSilentRequirement]). */
    fun displayableIssues(issues: List<FieldIssue>): List<FieldIssue> =
        issues.filterNot { it.code.isSilentRequirement }

    // Internal predicates.

    /**
     * Consistency between the connection mode and the address scheme. This is a form-level check,
     * not something discovered at submit time.
     *
     * It only judges the case where the user wrote a scheme **explicitly**: with no scheme, the
     * address policy supplies one from the current mode and the result is always compatible.
     * The other half - an `http://` address under a mode that requires encryption - is reported
     * by [RelayEndpointPolicy] as `cleartext_not_allowed`. This covers what that cannot see:
     * an encrypted address under a mode that declares cleartext.
     */
    private fun hasSchemeModeMismatch(
        endpoint: String,
        securityMode: RelayConnectionSecurityMode,
    ): Boolean {
        if (securityMode != RelayConnectionSecurityMode.LocalHttp &&
            securityMode != RelayConnectionSecurityMode.PrivateVpn
        ) {
            return false
        }
        return endpoint.lowercase().startsWith("https://")
    }

    /** Which piece of credential material actually crossed the line on a cleartext transport.
     *  Attributed in layout order so the user gets something they can tap. */
    private fun cleartextOffendingField(draft: RelayFormDraft, hasCredential: Boolean): Field {
        if (draft.authMode.requiresCredential || hasCredential) return Field.ApiKey
        if (draft.headers.any { RelayEndpointPolicy.isSensitiveName(it.key) }) return Field.Headers
        return Field.QueryParams
    }

    private fun reasonOf(error: Throwable): String =
        (error as? ProviderServiceError.InvalidConfiguration)?.detail ?: "invalid_url"
}

/**
 * The form draft. The add page and the edit sheet each run their own state machine, but both
 * collapse their state into this one draft before handing it to validation.
 */
data class RelayFormDraft(
    val endpoint: String = "",
    val apiKey: String = "",
    val authMode: RelayAuthMode = RelayAuthMode.Auto,
    val securityMode: RelayConnectionSecurityMode = RelayConnectionSecurityMode.RemoteHttps,
    val transport: RelayTransport = RelayTransport.Auto,
    val modelID: String = "",
    val headers: List<RelayKeyValue> = emptyList(),
    val queryParams: List<RelayKeyValue> = emptyList(),
    /** A non-empty key is already stored in the keystore. Always `false` for Create, which has
     *  no keystore yet. */
    val hasSavedCredential: Boolean = false,
) {
    /**
     * The configuration handed to [RelayEndpointPolicy]. It carries only the fields the security
     * boundary cares about, rather than assembling a shadow object that is almost the stored
     * configuration.
     */
    fun requestedConfigForPolicy(): RelayRequestedConfig = RelayRequestedConfig(
        transport = transport,
        authMode = authMode,
        securityMode = securityMode,
        headers = headers.ifEmpty { null },
        queryParams = queryParams.ifEmpty { null },
    )
}

/**
 * Opens a form from an already-stored custom-endpoint configuration. Both the edit sheet and the
 * add page's "prefill from a template" path go through here, so the protocol fields in the draft
 * always come from the same place as the stored ones.
 */
fun RelayFormDraft(
    requested: RelayRequestedConfig?,
    endpoint: String,
    apiKey: String = "",
    hasSavedCredential: Boolean = false,
): RelayFormDraft = RelayFormDraft(
    endpoint = endpoint,
    apiKey = apiKey,
    authMode = requested?.authMode ?: RelayAuthMode.Auto,
    securityMode = requested?.securityMode ?: RelayConnectionSecurityMode.RemoteHttps,
    transport = requested?.transport ?: RelayTransport.Auto,
    modelID = requested?.modelID.orEmpty(),
    headers = requested?.headers.orEmpty(),
    queryParams = requested?.queryParams.orEmpty(),
    hasSavedCredential = hasSavedCredential,
)
